src/Polysemy/Internal/Effunctor.hs

```diff
@@ -0,0 +1,27 @@
+module Polysemy.Internal.Effunctor where
+
+import Data.Kind
+import Polysemy.Internal
+import Polysemy.Internal.Kind
+
+-- | Effects that are \"functors\" over their argument.
+--
+-- TODO: add better docs
+class Effunctor (e1 :: Type -> Effect) where
+  effmap
+    :: forall r a b c
+     . Member (e1 b) r
+    => (a -> b)
+    -> Sem (e1 a ': r) c
+    -> Sem r c
```

Comment:
I would expect this to be `Sem (e1 b ': r) c`

Reply:
I mirrored
```haskell
runOuputAsAnother :: Member (Output o2) r => (o1 -> o2) -> Sem (Output o1 ': r) a -> Sem r a
```
from [165](https://github.com/polysemy-research/polysemy/issues/165#issuecomment-509036665) for the type. What are the pros/cons of explicitly putting it in the front? (type inference/error messages/internal machinery/performance issues?)

Reply:
Good point. That's a very motivating example! Explicitly putting it at the front would mean we can't solve the [trouble with typed errors](https://www.parsonsmatt.org/2018/11/03/trouble_with_typed_errors.html)

Reply:
I seem to often write types like

```haskell
data Action r = forall s . Action
  { startingValue :: s
  , action        :: Int -> Sem (State s ': r) ()
  }

printAction :: Member Trace r => Action r
interactiveSetAction :: Member (Lift IO) r => Action r
```

And then I want `Action r` to be a `Semigroup` which runs both actions with their respective state:

```haskell
instance Semigroup (Action r) where
  (Action s a) <> (Action s' a') = Action (s, s') $ \value -> do
    mapStateFirst' (a value)
    mapStateSecond' (a' value)

mapStateFirst' :: Sem (State a ': r) x -> Sem (State (a, b) ': r) x
mapStateSecond' :: Sem (State b ': r) x -> Sem (State (a, b) ': r) x
```

For this case, these functions won't work:
```haskell
mapStateFirst :: Member (State (a, b)) r => Sem (State a ': r) x -> Sem r x
mapStateSecond :: Member (State (a, b)) r => Sem (State b ': r) x -> Sem r x
```

Failing with
<details>

```
    • Occurs check: cannot construct the infinite type:
        r ~ State (s, s1) : r
      Expected type: Sem (State (s, s1) : r) ()
        Actual type: Sem r ()
    • In a stmt of a 'do' block: mapStateFirst (a value)
      In the expression:
        do mapStateFirst (a value)
           mapStateSecond (a' value)
      In the second argument of ‘($)’, namely
        ‘\ value
           -> do mapStateFirst (a value)
                 mapStateSecond (a' value)’
    • Relevant bindings include
        a' :: Int -> Sem (State s1 : r) () (bound at tests/Main.hs:29:30)
        s' :: s1 (bound at tests/Main.hs:29:27)
        a :: Int -> Sem (State s : r) () (bound at tests/Main.hs:29:13)
        s :: s (bound at tests/Main.hs:29:11)
        (<>) :: Action r -> Action r -> Action r
          (bound at tests/Main.hs:29:16)
   |
30 |     mapStateFirst (a value)
   |     ^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

I'm not sure if I'm designing something wrong here (should I not use `Semigroup`?), but if not, both of these function variants seem to have their use.

In my code I currently even use both variants. Usage is [here](https://github.com/Infinisil/arvy/blob/64f966cbd3afe0b7b87f480c94987f001c9a1be2/eval/Evaluation.hs#L23-L67), implementation of them is [here](https://github.com/Infinisil/arvy/blob/64f966cbd3afe0b7b87f480c94987f001c9a1be2/eval/Utils.hs#L38-L89).

Reply:
I haven't gone too deep into your example, but why not write `action` as `Member (State s) r => Int -> Sem r ()` instead?

Reply:
He later tries to map `State` in this action, which requires him to remove original and add a modified one - `Member` constraint is not enough for that.